src/net_processing.cpp

```diff
@@ -3262,6 +3290,55 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+            // This is an outbound peer subject to disconnection if their chain
+            // lags behind ours (note: if their chain has more work than ours,
+            // we should sync to it, unless it's invalid, in which case we
+            // should find that out and disconnect from them elsewhere).
+            if (state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= chainActive.Tip()->nChainWork) {
+                if (state.nHeadersChainTimeout != 0) {
+                    state.nHeadersChainTimeout = 0;
+                    state.header_with_required_work = nullptr;
+                    state.fSentGetHeadersToCheckChainSync = false;
+                }
+            } else if (state.nHeadersChainTimeout == 0 || (state.header_with_required_work != nullptr && state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= state.header_with_required_work->nChainWork)) {
+                // Our best block known by this peer is behind our tip, and we're either noticing
+                // that for the first time, OR this peer was able to catch up to some earlier point
+                // where we checked against our tip.
+                // Either way, set a new timeout based on current tip.
+                state.nHeadersChainTimeout = nNow + CHAIN_SYNC_TIMEOUT;
+                state.header_with_required_work = chainActive.Tip();
+                state.fSentGetHeadersToCheckChainSync = false;
+            } else if (state.nHeadersChainTimeout > 0 && nNow > state.nHeadersChainTimeout) {
+                // No evidence yet that our peer has synced to a chain with work equal to that
+                // of our tip, when we first detected it was behind.  Send a single getheaders
+                // message to give the peer a chance to update us.
+                if (state.fSentGetHeadersToCheckChainSync) {
+                    // They've run out of time to catch up!
+                    LogPrintf("Disconnecting outbound peer %d for old chain, best known block = %s\n", state.pindexBestKnownBlock->GetBlockHash().ToString());
+                    pto->fDisconnect = true;
+                } else {
+                    const CBlockIndex *pindexStart = state.header_with_required_work;
+                    assert (pindexStart != nullptr);
+                    if (pindexStart->pprev != nullptr) {
+                        pindexStart = pindexStart->pprev;
+                    }
+                    LogPrint(BCLog::NET, "sending getheaders (%d) to outbound peer=%d to verify chain work (current best known block:%s)\n", pindexStart->nHeight, pto->GetId(), state.pindexBestKnownBlock->GetBlockHash().ToString());
+                    connman->PushMessage(pto, msgMaker.Make(NetMsgType::GETHEADERS, chainActive.GetLocator(state.header_with_required_work->pprev), uint256()));
```

Comment:
Would probably be clearer to reference `pindexStart` here instead of `state.header_with_required_work->pprev`.

Reply:
...or just strip out the `pindexStart` stuff altogether because it seems like that variable is unnecessary.